cmd/systray/logo.go

```diff
@@ -17,113 +18,190 @@ import (
 	"github.com/fogleman/gg"
 )
 
-// tsLogo represents the state of the 3x3 dot grid in the Tailscale logo.
-// A 0 represents a gray dot, any other value is a white dot.
-type tsLogo [9]byte
+// tsLogo represents the Tailscale logo displayed as the systray icon.
+type tsLogo struct {
+	// dots represents the state of the 3x3 dot grid in the logo.
+	// A 0 represents a gray dot, any other value is a white dot.
+	dots [9]byte
+
+	// dotMask returns an image mask to be used when rendering the logo dots.
+	dotMask func(dc *gg.Context, borderUnits int, radius int) *image.Alpha
+
+	// overlay is called after the dots are rendered to draw an additional overlay.
+	overlay func(dc *gg.Context, borderUnits int, radius int)
+}
 
 var (
 	// disconnected is all gray dots
-	disconnected = tsLogo{
+	disconnected = tsLogo{dots: [9]byte{
 		0, 0, 0,
 		0, 0, 0,
 		0, 0, 0,
-	}
+	}}
 
 	// connected is the normal Tailscale logo
-	connected = tsLogo{
+	connected = tsLogo{dots: [9]byte{
 		0, 0, 0,
 		1, 1, 1,
 		0, 1, 0,
-	}
+	}}
 
 	// loading is a special tsLogo value that is not meant to be rendered directly,
 	// but indicates that the loading animation should be shown.
-	loading = tsLogo{'l', 'o', 'a', 'd', 'i', 'n', 'g'}
+	loading = tsLogo{dots: [9]byte{'l', 'o', 'a', 'd', 'i', 'n', 'g'}}
 
 	// loadingIcons are shown in sequence as an animated loading icon.
 	loadingLogos = []tsLogo{
-		{
+		{dots: [9]byte{
 			0, 1, 1,
 			1, 0, 1,
 			0, 0, 1,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 1, 1,
 			0, 0, 1,
 			0, 1, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 1, 1,
 			0, 0, 0,
 			0, 0, 1,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 1,
 			0, 1, 0,
 			0, 0, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 1, 0,
 			0, 0, 0,
 			0, 0, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			0, 0, 1,
 			0, 0, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			0, 0, 0,
 			0, 0, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 1,
 			0, 0, 0,
 			0, 0, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			0, 0, 0,
 			1, 0, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			0, 0, 0,
 			1, 1, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			1, 0, 0,
 			1, 1, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			1, 1, 0,
 			0, 1, 0,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			1, 1, 0,
 			0, 1, 1,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 0, 0,
 			1, 1, 1,
 			0, 0, 1,
-		},
-		{
+		}},
+		{dots: [9]byte{
 			0, 1, 0,
 			0, 1, 1,
 			1, 0, 1,
+		}},
+	}
+
+	// exitNodeOnline is the Tailscale logo with an additional arrow overlay in the corner.
+	exitNodeOnline = tsLogo{
```

Comment:
these really feel like different named states a logo could be in at this point, rather than just a dot arrangement. in fact, it's definitely not a dot arrangement for the loading case. maybe rather than being package-level vars they can be an enum, and the drawing code just draws one of them based on the enum rather than passing around 9 bytes + 2 pointers every time?

Reply:
maybe?  Given that we have to setup the mask before drawing the dots, and then draw the overlay afterward, I'm not sure that the result would be any more legible.  Fortunately, state changes don't happen terribly often, so I'm not terribly worried about copying the data around, and think the legibility of keeping the masking and overlay logic together is worth it.  I think there are at least a couple more variants that other clients have that I'll need to add later.

Reply:
alright, if you've considered it 👍